### DISCUSSION TITLE
Extended Field Array Example

### DISCUSSION THREAD

**Discussion by 6thfdwp** [react-hook-form/react-hook-form#3644] [Q&A]
First thanks for the great work for such a clean form solution! 
My question is more about understanding about the library and the right way to handle an array of field object.

I stumbled into one old chat in Spectrum which is very similar to my use case:
https://spectrum.chat/react-hook-form/help/resetting-subform-while-retaining-the-mainform~6d17bdb1-70a7-465a-9f2e-17ae5cf8367f
and trying to extend the field array example to achieve that. I tried to simplify it to its minimal scenario, hopefully easy to understand. Basically we want to collect such form data: (imagine creating a  meeting room)
```js
useForm({
    meeting: "",
    users: [{firstName: '', lastName: '' }]
  });
```
Here is the sandbox url that we could see the result: 
https://codesandbox.io/s/react-hook-form-v6-controller-append-prepepend-insert-forked-wnqyx?file=/src/sub-form-with-field-arry.js

Compared to the field array, the main difference is to have `UserSubForm` to provide initial data for each added user, seems it's problematic to nest the `useForm`.  when we do not edit these added users, the data looks fine 
![image](https://user-images.githubusercontent.com/2919741/101724211-c289a380-3af9-11eb-8dfd-31713daf2c0b.png)

But let's say user tries to edit in the user list, when submitting, the user list got reset to initial even though the printed form data looks as expected as well 
![image](https://user-images.githubusercontent.com/2919741/101724584-a89c9080-3afa-11eb-80c9-6345b632a103.png)

Maybe we should not nested `useForm` in `UserSubForm`, just manage its own state and pass it in the callback when hitting `Add`? 
Would be good if you could explain a bit more about how the user list got reverted back, (noticed the whole form got re-rendered twice, the second time the form data changed back to initial)

Thanks in advance!

**Comment by bluebill1049** [comment]
hey @6thfdwp is this what you after? https://codesandbox.io/s/react-hook-form-usefieldarray-nested-arrays-x7btr

**Comment by 6thfdwp** [comment]
Hey @bluebill1049  thanks for taking a look. 

Not really, it's just one level of array if you look at the structure 
```js
{
    meeting: "",
    users: [{firstName: '', lastName: '' }]
  }
```

I think the confusing part is in my scenario, user would first fill in first and last name, click add and use the input values to create the item in the list. That's why you see extra two inputs compared to your example, these two inputs are not actually part of the main form, just to provide initial value to fill in inputs for each generated item.
![image](https://user-images.githubusercontent.com/2919741/101762221-de586e00-3b28-11eb-9d0d-e696bfe6de3e.png)

If it is just user directly clicks the 'Add' button, and create the inputs for each item with default value, (and user would further update there). this would work as expected.

Hopefully I make myself clear.

**Comment by 6thfdwp** [comment]
OK this is my intent, forked from your simple useFieldArray example 
https://codesandbox.io/s/react-hook-form-usefieldarray-forked-bfdyi?file=/src/index.js

This is actually work, use `fields` from `useFieldArray` hook to drive the list rendering.  After submitting, the updated inputs value in the list stay the same 

Initially I followed this one from the official doc
https://codesandbox.io/s/react-hook-form-v6-controller-append-prepepend-insert-7clz7?file=/src/index.js
It does not the `useFieldArray` hook, instead has its own state to drive the list rendering, but inputs value in the list also not reverted to initial value upon submitting
```js
  const [data, setData] = useState([]);
```
